File: second_year/ICP/src/simulation.cpp
/**
 * Implementation of the logic simulation
 * @author Patrik Holop
 */


#include "simulation.h"


/**
 * Check, whether current block is connected to any cycle using DFS
 * @brief Simulation::check_cycles_block
 * @param block_ip block to be checked
 * @param path Current path (for backtracking)
 * @return True is block is connected to cycle, otherwise false
 */
bool Simulation::check_cycles_block(int block_ip, std::vector<int> path) {
    std::cout << "Considering: " << block_ip << ", PATH: "; for (int id: path) std::cout << id; std::cout << std::endl;
    // find the correct block
    Block *b = NULL;
    for (Block *bl : this->blocks) {
        if (block_ip == bl->local_id)
            b = bl;
    }
    // for each connection of the block
    for (Port *p : b->outputPorts) {
        for (Connection *c : p->connections) {
            // if the other block is in already taken path, it's a cycle
            bool find = false;
            for (int id : path) {
                if (id == c->outputPort->loc) {
                    find = true;
                }
            }
            if (find) {
                std::cout << "CYCLE" << std::endl;
                return true;
            }
            path.push_back(c->outputPort->loc);
            // otherwise add it to the path and check the next block
            bool sub_cycle = check_cycles_block(c->outputPort->loc, path);
            if (sub_cycle)
                return true;
            path.pop_back();
        }
    }
    std::cout << "no cycle" << std::endl;
    return false;
}


/**
 * Checks, whether there's at least one cycle in the current scheme
 * @brief Simulation::check_cycles
 * @return True if the scheme contains cycles, otherwise false
 */
bool Simulation::check_cycles() {
    std::vector<int> path;
    // neither of the bloks can contain cycle to be the valid scheme
    for (Block *b : this->blocks) {
        std::cout << "Check cycles for: " << b->local_id << std::endl;
        if (check_cycles_block(b->local_id, path))
            return true;
    }
    return false;
}


/**
 * Sort the block in correct execution order to correctly update the whole scheme.
 * @brief Simulation::sort_blocks
 */
void Simulation::sort_blocks() {
    // valid sorted order
    std::vector<int> ids;
    while (ids.size() != this->blocks.size()) {
        for (Block *b: this->blocks) {
            // propagate all blocks that have no input values
            if (b->inputPorts.size() == 0) {
                bool find = false;
                for (int id : ids) {
                    if (id == b->local_id) {
                        find = true;
                    }
                }
                if (!find) {
                    ids.push_back(b->local_id);
                }
            } else {
                // to add the block, check whether all of it's ancestors are
                // already updated
                bool add = true;                
                std::cout << "CHECKING: " << b->local_id << std::endl;
                for (Port *p: b->inputPorts) {
                    std::cout << "connections to check: " << p->connections.size() << std::endl;
                    for (Connection *c: p->connections) {
                        bool find = false;
                        for (int id : ids) {
                            std::cout << "COMPARE: " << id << " == " << c->inputPort->loc << std::endl;
                            if (id == c->inputPort->loc)
                                find = true;
                        }
                        if (!find) {
                            add = false;
                        } else {
                            std::cout << "CONNECTION OK" << std::endl;
                        }
                    }
                }
                if (add) {
                    std::cout << "ADDING: " << b->local_id << std::endl;
                    bool find = false;
                    for (int id : ids) {
                        if (id == b->local_id)
                            find = true;
                    }
                    if (!find) {
                        ids.push_back(b->local_id);
                    } else {
                        std::cout << "FOUND";
                    }
                }
            }
        }
        for (auto b: ids) {
            std::cout << b << " ";
        }
        std::cout << "IDS: " << ids.size() << " BLOCKS: " << this->blocks.size() << std::endl;
    }

    for (auto f: ids) {
        std::cout << "F: " << f << std::endl;
    }
    // generate new logic block array from the id vector
    std::vector <Block *> new_main_block_array;
    for (float id: ids) {
        Block *to_add = NULL;
        for (Block *b: this->blocks) {
            if (b->local_id == id) {
                to_add = b;
            }
        }
        new_main_block_array.push_back(to_add);
    }
    std::cout << "ORDER" << std::endl;
    for (Block *b : new_main_block_array) {
        std::cout << b->local_id << ", ";
    }
    std::cout<< "\nEND ORDER" << std::endl;
    this->blocks = new_main_block_array;
}

File: second_year/IJA/src/main/Simulation.java
/*
 * This class is responsible for simulation management
 * Provides functions to step or run the whole simulation
 * @author Patrik Holop
 */
package main;

import javafx.animation.KeyFrame;
import javafx.animation.Timeline;
import javafx.scene.control.Alert;
import javafx.scene.control.TextInputDialog;
import javafx.scene.paint.Color;
import javafx.scene.shape.Circle;
import javafx.util.Duration;
import logic.base.Block;
import logic.base.Port;
import logic.connections.Connection;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.Map;
import java.util.Optional;


/**
 * This class is responsible for simulation management
 * Provides functions to step or run the whole simulation
 * @author Patrik Holop
 */
public class Simulation {

    /**
     * Gets all unassigned values for ports
     */
    static void get_unassigned_values() {
        for (Block b: Main.main.blocks) {
            for (Port p: b.inputPorts) {
                if (p.connections == null || p.connections.size() == 0) {
                    boolean stop = false;
                    while (!stop) {
                        stop = true;
                        Iterator it = p.get_value().entrySet().iterator();
                        while (it.hasNext()) {
                            Map.Entry pair = (Map.Entry) it.next();
                            // if the value was unassigned
                            if (pair.getValue() == null) {
                                // get the value from user
                                TextInputDialog dialog = new TextInputDialog("1.0");
                                dialog.setTitle(b.name + " " + p.name + " : " + pair.getKey());
                                dialog.setHeaderText("Please enter valid value");
                                dialog.setContentText("Value:");
                                Optional<String> value = dialog.showAndWait();
                                try {
                                    p.get_value().put(pair.getKey().toString(), Double.parseDouble(value.get()));
                                } catch (NumberFormatException e) {
                                    // pass when wrong number was assigned
                                    stop = false;
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    /**
     * Sorts blocks in the main application in order of execution
     */
    static void sort_blocks() {
        ArrayList<Float> ids = new ArrayList<>();

        // while not all blocks are sorted
        while (ids.size() != Main.main.blocks.size()) {
            for (Block b: Main.main.blocks) {
                // if the block has no input ports, it might be inserted
                if (b.inputPorts == null || b.inputPorts.size() == 0) {
                    if (!ids.contains(b.local_id)) {
                        ids.add(b.local_id);
                    }
                }
                else {
                    // if the block has input ports, each of them needs to be part
                    // of already added block to solve dependence
                    boolean add = true;
                    for (Port p: b.inputPorts) {
                        for (Connection c: p.connections) {
                            if (!ids.contains(c.inputPort.loc)) {
                                add = false;
                            }
                        }
                    }
                    if (add) {
                        if (!ids.contains(b.local_id)) {
                            ids.add(b.local_id);
                        }
                    }
                }
            }
        }
        // new sorted array of blocks
        ArrayList<Block> new_main_block_array = new ArrayList<>();
        for (Float id : ids) {
            // find correct block and add it to sorted array
            Block to_add = null;
            for (Block b : Main.main.blocks) {
                if (b.local_id == id) {
                    to_add = b;
                }
            }
            new_main_block_array.add(to_add);
        }
        // reassign the arrays
        Main.main.blocks = new_main_block_array;
    }

    /**
     * This method runs the whole simulation.
     * Before start checks, whether exists any cycle in created scheme
     */
    public static void run_simulation() {

        // find cycles
        boolean has_cycle = SchemeChecker.containsCycles();

        if (has_cycle == true) {
            // error if found
            Alert alert = new Alert(Alert.AlertType.ERROR);
            alert.setTitle("Block Editor Error");
            alert.setHeaderText("Cycle detected");
            alert.setContentText("In current scheme was detected cycle!\nPlease make necessary changes and try again.");
            alert.showAndWait();
        }
        else {
            // sort blocks
            sort_blocks();

            // get missing values
            get_unassigned_values();

            for (Block b: Main.main.blocks) {
                if (b.active == true) {
                    b.active = false;
                }
            }
            // run simulation with delay on every block
            Timeline timeline = new Timeline(new KeyFrame(Duration.millis(1000), ae -> update_step_scheme()));
            timeline.setCycleCount(Main.main.blocks.size());
            timeline.play();
        }
    }

    /**
     * This method updates one block with it's connections and chooses the next block to be active
     */
    public static void update_step_scheme() {
        sort_blocks();

        // find block objects based on activity
        Block to_update = null;
        int pos = 0;
        int index = 0;
        for (Block next : Main.main.blocks) {
            if (next.active) {
                to_update = next;
                pos = index;
            }
            index++;
        }
        // if no object found (first run), set the first to be active
        if (Main.main.blocks.size() == 0)
            return ;
        if (to_update == null) {
            to_update = Main.main.blocks.get(0);
        }

        // remove previous activity circle marker
        Main.main.all_drawable_objects.getChildren().remove(Main.main.active_circle);
        // draw green circle marking the active block
        Main.main.active_circle = new Circle((int)to_update.group.getTranslateX(), (int)to_update.group.getTranslateY(), 5, Color.GREEN);
        Main.main.all_drawable_objects.getChildren().add(Main.main.active_circle);


        // check if no value is needed to add by user
        for (Port p: to_update.inputPorts) {
            if (p.connections == null || p.connections.size() == 0) {
                boolean stop = false;
                while (!stop) {
                    stop = true;
                    Iterator it = p.get_value().entrySet().iterator();
                    while (it.hasNext()) {
                        Map.Entry pair = (Map.Entry) it.next();
                        // if the value was unassigned
                        if (pair.getValue() == null) {
                            // get the value from user
                            TextInputDialog dialog = new TextInputDialog("1.0");
                            dialog.setTitle(to_update.name + " " + p.name + " : " + pair.getKey());
                            dialog.setHeaderText("Please enter valid value");
                            dialog.setContentText("Value:");
                            Optional<String> value = dialog.showAndWait();
                            try {
                                p.get_value().put(pair.getKey().toString(), Double.parseDouble(value.get()));
                            } catch (NumberFormatException e) {
                                // pass when wrong number was assigned
                                stop = false;
                            }
                        }
                    }
                }
            }
        }

        // update_tooltip the block
        to_update.update();

        // update_tooltip all outer connection from the block
        for (Connection next : Main.main.connections) {
            if (next != null) {
                if (next.inputPort.loc == to_update.local_id) {
                    next.update_tooltip();
                    Block output_block= null;
                    double loc = next.outputPort.loc;
                    for (Block output: Main.main.blocks) {
                        if (loc == output.local_id) {
                            output.update();
                        }
                    }
                }
            }
        }

        // set the next active block
        if (pos+1 < Main.main.blocks.size()) {
            to_update.active = false;
            Main.main.blocks.get(pos+1).active = true;
        }
        else {
            to_update.active = false;
            Main.main.blocks.get(0).active = true;
        }
    }
}
